data/beaconrestapi/src/main/java/tech/pegasys/artemis/beaconrestapi/beaconhandlers/BeaconBlockHandler.java

```diff
@@ -15,60 +15,79 @@
 import tech.pegasys.artemis.datastructures.blocks.BeaconBlock;
+import tech.pegasys.artemis.datastructures.blocks.SignedBeaconBlock;
 import tech.pegasys.artemis.storage.ChainStorageClient;
+import tech.pegasys.artemis.storage.HistoricalChainData;
 
 public class BeaconBlockHandler implements BeaconRestApiHandler {
 
   private final ChainStorageClient client;
+  private final HistoricalChainData historicalChainData;
 
-  public BeaconBlockHandler(ChainStorageClient client) {
+  public BeaconBlockHandler(ChainStorageClient client, EventBus eventBus) {
     this.client = client;
+    this.historicalChainData = new HistoricalChainData(eventBus);
   }
 
   @Override
   public String getPath() {
-    return "/beacon/block/";
+    return "/beacon/block";
   }
 
   @Override
   public Object handleRequest(RequestParams param) {
     Map<String, List<String>> queryParamMap = param.getQueryParamMap();
-    Map<String, Object> jsonObject = new HashMap<>();
-    BeaconBlock block;
-    Bytes32 blockRoot;
     if (queryParamMap.containsKey("root")) {
       Bytes32 root = Bytes32.fromHexString(param.getQueryParam("root"));
       return client.getStore() != null ? client.getStore().getBlock(root) : null;
-    } else if (queryParamMap.containsKey("epoch")) {
-      UnsignedLong epoch = UnsignedLong.valueOf(param.getQueryParam("epoch"));
-      Optional<Bytes32> blockRootAtSlot =
-          client.getBlockRootBySlot(compute_start_slot_at_epoch(epoch));
-      blockRoot = blockRootAtSlot.orElse(null);
-      block =
-          client.getStore() != null && blockRootAtSlot.isPresent()
-              ? client.getStore().getBlock(blockRootAtSlot.get())
-              : null;
+    }
+
+    UnsignedLong slot;
+    if (queryParamMap.containsKey("epoch")) {
+      slot = compute_start_slot_at_epoch(UnsignedLong.valueOf(param.getQueryParam("epoch")));
     } else if (queryParamMap.containsKey("slot")) {
-      UnsignedLong slot = UnsignedLong.valueOf(param.getQueryParam("slot"));
-      Optional<Bytes32> blockRootAtSlot = client.getBlockRootBySlot(slot);
-      blockRoot = blockRootAtSlot.orElse(null);
-      block =
-          client.getStore() != null && blockRootAtSlot.isPresent()
-              ? client.getStore().getBlock(blockRootAtSlot.get())
-              : null;
+      slot = UnsignedLong.valueOf(param.getQueryParam("slot"));
     } else {
       return null;
     }
-    jsonObject.put("block", block);
-    jsonObject.put("blockRoot", blockRoot.toHexString());
+
+    Optional<Pair<BeaconBlock, Bytes32>> blockAndRoot = getBlockBySlot(slot);
+    if (blockAndRoot.isEmpty()) {
+      return null;
+    }
+
+    Map<String, Object> jsonObject = new HashMap<>();
+    jsonObject.put("blockRoot", blockAndRoot.get().getRight().toHexString());
+    jsonObject.put("block", blockAndRoot.get().getLeft());
+
     return jsonObject;
   }
+
+  private Optional<Pair<BeaconBlock, Bytes32>> getBlockBySlot(UnsignedLong slot) {
```

Comment:
This can just return `Optional<BeaconBlock>` since the root is just calculated by calling `block.hash_tree_root()` anyway.

Reply:
I only did that because if we're able to find the block in memory, we already access it using its root. I guess assuming that the hash tree roots will be cached going forward, this is unnecessary.

Reply:
Done.